### PR TITLE
Add support for Snowflake create/alter SQL and js UDF

### PR DESCRIPTION
**Pull request by chwiese**
### Brief summary of the change made

Adds support for SQL and JavaScript [CREATE FUNCTION](https://docs.snowflake.com/en/sql-reference/sql/create-function.html) as well as [ALTER FUNCTION](https://docs.snowflake.com/en/sql-reference/sql/alter-function.html) statements in Snowflake. IMHO, the `CREATE FUNCTION` syntax for Java UDFs is sufficiently different that it should have its own statement.

### Are there any other side effects of this change that we should be aware of?

No.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `python test/generate_parse_fixture_yml.py` or by running `tox` locally).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/dialects/dialect_snowflake.py [hunk 1]
```diff
@@ -982,7 +983,7 @@ class CreateProcedureStatementSegment(BaseSegment):
     https://docs.snowflake.com/en/sql-reference/sql/create-procedure.html
     """
 
-    type = "create_clone_statement"
```
Comment (barrywhart) [inline src/sqlfluff/dialects/dialect_snowflake.py:985]: Is this a bug fix? It seems unrelated to the rest of the PR.

**Review by barrywhart — COMMENTED** [review]
I think this looks good. One question.

File: src/sqlfluff/dialects/dialect_snowflake.py (hunk 1 above)
Comment (chwiese) [reply src/sqlfluff/dialects/dialect_snowflake.py:985]: Yes, kind of a drive by bug fix. I assume that the `CreateCloneSegment` was copied to create this segment (same as I copied this segment), but the author forgot to change this value.

**Review by barrywhart — COMMENTED** [review]
Looks good!

**Review by barrywhart — APPROVED** [review]
Looks great!
